network: Add support specify the outgoing interface for FDB entries

ssahani opened the pull request:
Closes https://github.com/systemd/systemd/issues/17247

poettering commented on src/network/networkd-fdb.c:
```diff
@@ -392,3 +399,46 @@ int config_parse_fdb_ntf_flags(
+        if (ifindex < 0) {
+                log_error_errno(r, "Could not resolve Interface='%s': %m", rvalue);
+                return 0;
+        }
```
this doesn't look right: parsing this at the time the .network files are read might be too late. this should store the ifname as text and resolve when the fdb record is actually installed, so that this works for all network interfaces created between the time networkd reads the .network files and then finally acts on them.

yuwata replied:
Yeah, but we already use resolve_ifname() at many config parsers. We need to fix them at some point...

ssahani replied:
How does it fix ? altname or rename

yuwata replied:
Maybe, store the ifname as is in conf parsers, and then resolve it in configure_xxx().

ssahani replied:
Is it the timing ?

yuwata replied:
Yeah, at that time when we parse the .network files, the corresponding netdevs may not exist yet.

yuwata replied:
But to handle such scenario is not simple. Maybe keep this as is, and leave to a later PR is ok at this time, I think. @poettering WDYT?

ssahani replied:
yes it's strange how to handle this.

poettering replied:
if this is already a pre-existing pattern it might be Ok to merge this with the current logic, but this really should be fixed I guess. I don#t think the fix is going to be difficult: it just means storing the string the user specified and then at the time of applying things looking for the right ifindex for it. Given that networkd keeps tracks of all ifaces anyway that should be pretty easy.

poettering reviewed (changes requested):
This looks pretty OK as it stands now, to me, I'll leave it to @yuwata to merge that.

As mentioned, I think the Manager object should be improved so that we don't have to go to rtnl to look up an ifname (or ifaltname) to an ifindex. after all we have all the data in memory anyway for that, but not indexed in any hash table. i.e. we should add thta hash table so that that just works.

ssahani commented on src/network/networkd-fdb.c:
```diff
@@ -146,6 +148,13 @@ static int fdb_entry_configure(Link *link, FdbEntry *fdb_entry) {
                         return log_link_error_errno(link, r, "Could not append NDA_VLAN attribute: %m");
         }
 
+        ifindex = resolve_ifname(NULL, fdb_entry->ifname);
```
Sorry for late reply. Yes but I think rather keeping extra hashmap we can introduce a function `manager_find_link_by_ifname` to do the same. Please see below.

ssahani commented:
Updated thanks for the review ptal.

poettering commented on src/network/networkd-link.c:
```diff
@@ -454,6 +454,24 @@ static int link_new(Manager *manager, sd_netlink_message *message, Link **ret) {
+                        if (r < 0)
+                                return r;
+                }
+        }
```
the names need to be removed again, when the link goes away. And when the link is renamed/altnames added/removed the hash table also needs updating

ssahani replied:
Ahh right. updated thanks.

poettering commented:
i like where this is going, but the hashtable needs a bit more love to always stay correctly up-to-date, see above

poettering commented on src/network/networkd-link.c:
```diff
@@ -1692,6 +1726,15 @@ static void link_detach_from_manager(Link *link) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove(link->manager->links_by_name, *n);
```
as a safety precaution I think we should use hashmap_remove_value() here. The only difference is that it also takes both key *and* value when removing, and only removes the entry when the key actually points to the indicated value. i.e use 

```c
hashmap_remove_value(link->manager->links_by_name, *n, link);
```

This might be useful in case the hash table ever gets out of sync, because of some bug.

poettering commented:
looks good, but link_update() needs to be adjusted too, to update the new hash table whenever IFLA_IFNAME or IFLA_ALT_IFNAME change

ssahani commented:
> looks good, but link_update() needs to be adjusted too, to update the new hash table whenever IFLA_IFNAME or IFLA_ALT_IFNAME change

Yes right  I removed and added them back in case they changed in `link_update` thanks. But I tried to test it adding a altname while `networkd` is running, the kernel is not pushing the changes or may be we need to add some other flag in the monitor. Updated ptal.

poettering commented on src/network/networkd-link.c:
```diff
@@ -2638,11 +2681,23 @@ int link_update(Link *link, sd_netlink_message *m) {
                         return r;
+
+                hashmap_remove_value(link->manager->links_by_name, link->ifname, link);
+                hashmap_put(link->manager->links_by_name, ifname, link);
```
oom not handled. maybe better move this into link_drop() and link_add() and make the latter fail if this doesn't work?

ssahani replied:
OK

poettering commented on src/network/networkd-link.c:
```diff
@@ -2638,11 +2681,23 @@ int link_update(Link *link, sd_netlink_message *m) {
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
+
+                STRV_FOREACH(n, s)
+                        hashmap_put(link->manager->links_by_name, *n, link);
```
oom not checked.

i guess the negative effect of this is not too bad, hence might make sense, to just log about this and proceed, i.e. `log_warning_errno()`

ssahani commented:
Updated ptal thanks !

poettering commented:
hmm, but the hash table is not updated in link_add()/link_drop either here. What am I missing?

ssahani commented:
`link_add` calls `link_new` there it gets updated and link_drop it's getting updated https://github.com/systemd/systemd/pull/17248/commits/defbb4a65f7a5eacf2345707ae07fc9790df2297#diff-3530f9d105ef619fb06e6c8b7fff216475b94d37d9be2a1b673648a7d0080925R1749.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -454,6 +454,24 @@ static int link_new(Manager *manager, sd_netlink_message *message, Link **ret) {
+        if (r < 0)
+                return r;
+
+        if (link->alternative_names) {
```
This condition is not necessary.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -1712,6 +1746,14 @@ void link_drop(Link *link) {
 
         log_link_debug(link, "Link removed");
 
+        if (link->alternative_names) {
```
This condition is not necessary.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -1712,6 +1746,14 @@ void link_drop(Link *link) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
```
Why do you use `hashmap_remove_value()`? It seems `hashmap_remove()` is sufficient.

ssahani replied:
See https://github.com/systemd/systemd/pull/17248#discussion_r507828602

yuwata replied:
Thank you for the reference. OK, let's keep this as is.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -1712,6 +1746,14 @@ void link_drop(Link *link) {
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
+        }
+        hashmap_remove_value(link->manager->links_by_name, link->ifname, link);
```
Same here. Whey `hashmap_remove_value()`?

yuwata commented on src/network/networkd-link.c:
```diff
@@ -2641,8 +2683,22 @@ int link_update(Link *link, sd_netlink_message *m) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
```
`hashmap_remove()` seems sufficient.

yuwata commented on src/network/networkd-fdb.c:
```diff
@@ -407,3 +422,42 @@ int config_parse_fdb_ntf_flags(
+        if (r < 0)
+                return log_oom();
+
+        if (!ifname_valid(rvalue)) {
```
I think, this should accept alternative names. Please see `ifname_valid_full()`.

yuwata commented on src/network/networkd-fdb.c:
```diff
@@ -146,6 +148,19 @@ static int fdb_entry_configure(Link *link, FdbEntry *fdb_entry) {
+        if (fdb_entry->ifname) {
+                Link *l;
+
+                r = link_get_by_name(link->manager, fdb_entry->ifname, &l);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to resolve ifname %s: %m", fdb_entry->ifname);
```
If this fails, then skip to configure this entry. We should not configure incomplete entries.

yuwata reviewed (changes requested):
In networkd-link.c, the list of alternative names are updated in several places. So, let's introduce, say `update_links_by_name()`.

ssahani commented:
Updated thanks for the review @yuwata ptal.

ssahani commented:
Updated thanks for the review @yuwata ptal.

ssahani commented:
@poettering ptal

yuwata commented on man/systemd.network.xml:
```diff
@@ -2529,6 +2529,11 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>Interface=</varname></term>
```
Maybe, `OutgoingInterface=`? In `[RoutingPolicyRule]` we already have `OutgoingInterface=`. So, I prefer `OutgoingInterface=` for consistency.

yuwata commented on src/network/networkd-fdb.c:
```diff
@@ -146,6 +148,18 @@ static int fdb_entry_configure(Link *link, FdbEntry *fdb_entry) {
                         return log_link_error_errno(link, r, "Could not append NDA_VLAN attribute: %m");
         }
 
+        if (fdb_entry->ifname) {
+                Link *l;
+
+                r = link_get_by_name(link->manager, fdb_entry->ifname, &l);
+                if (r < 0)
+                        return log_link_error_errno(link, r, "Failed to resolve ifname %s: %m", fdb_entry->ifname);
+
+                r = sd_netlink_message_append_u32(req, NDA_IFINDEX, l->ifindex);
+                if (r < 0)
+                        return log_link_error_errno(link, r, "Could not append NDA_IFINDEX attribute: %m");
+        }
```
I think we should skip to configure the entry rather than return failure when the device does not exist yet.
And, configure the entry when the corresponding interface comes.

Such a mechanism should also be used for `MultipathRoute=`, and not sure but maybe `OutgoingInterface=` and `IncomingInterface=` in `[RoutingPolicyRule]`.

ssahani replied:
https://github.com/systemd/systemd/pull/17248#discussion_r510582939. Then why did you made this comment

yuwata commented on src/network/networkd-fdb.c:
```diff
@@ -407,3 +421,42 @@ int config_parse_fdb_ntf_flags(
+        r = fdb_entry_new_static(network, filename, section_line, &fdb_entry);
+        if (r < 0)
+                return log_oom();
+
```
Please unset if an empty string is assigned.
```
if (isempty(rvalue)) {
        ftb_entry->ifname = mfree(fdb_entry->ifname);
        TAKE_PTR(fdb_entry);
        return 0;
}
```

yuwata commented on src/network/networkd-fdb.h:
```diff
@@ -33,6 +33,7 @@ typedef struct FdbEntry {
 
         int family;
         uint16_t vlan_id;
+        char *ifname;
```
I prefer `outgoing_ifname`.

ssahani replied:
It does not matter

yuwata commented on src/network/networkd-link.c:
```diff
@@ -454,6 +454,18 @@ static int link_new(Manager *manager, sd_netlink_message *message, Link **ret) {
         if (r < 0)
                 return r;
 
+        r = hashmap_ensure_allocated(&manager->links_by_name, &string_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = hashmap_put(manager->links_by_name, link->ifname, link);
+        if (r < 0)
+                return r;
+
```
I think these lines should be moved to `link_update_by_name()`.

ssahani replied:
The allocation should he here .

yuwata commented on src/network/networkd-link.h:
```diff
@@ -201,6 +201,9 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(Link*, link_unref);
 
 int link_get(Manager *m, int ifindex, Link **ret);
+int link_get_by_name(const Manager *m, const char *ifname, Link **ret);
+int link_update_by_name(Link *link);
```
It seems `link_update_by_name()` can be static.

ssahani replied:
It may be used somewhere later. Let's keep it that way.

yuwata commented on src/network/networkd-manager.c:
```diff
@@ -872,6 +872,7 @@ void manager_free(Manager *m) {
         m->dirty_links = set_free_with_destructor(m->dirty_links, link_unref);
         m->links_requesting_uuid = set_free_with_destructor(m->links_requesting_uuid, link_unref);
         m->links = hashmap_free_with_destructor(m->links, link_unref);
+        m->links_by_name = hashmap_free(m->links_by_name);
```
This must be in the first commit.

yuwata commented on src/network/networkd-manager.h:
```diff
@@ -40,6 +40,7 @@ struct Manager {
         LinkAddressState address_state;
 
         Hashmap *links;
+        Hashmap *links_by_name;
```
This must be in the first commit.

yuwata reviewed (changes requested):
Please move the lines about `manager->links_by_name` to the first commit, and make the second commit contain only about fdb related changes.

ssahani commented:
Cc @bluca

poettering commented on src/network/networkd-link.c:
```diff
@@ -457,6 +457,18 @@ static int link_new(Manager *manager, sd_netlink_message *message, Link **ret) {
+
+        r = hashmap_put(manager->links_by_name, link->ifname, link);
+        if (r < 0)
+                return r;
```
maybe move all this also into link_update_by_name(), so that that function comprehensively adds all names to the hashmap both the main one and the alternative ones, and implicitly creates the hashmap before, if it is missing

poettering commented on src/network/networkd-link.c:
```diff
@@ -2714,8 +2765,18 @@ int link_update(Link *link, sd_netlink_message *m) {
+
+                r = link_update_by_name(link);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to add alternative name '%s' on network interface '%s', ignoring: %m", *n, link->ifname);
```
you are not actually ignoring here. I think ignoring is good, but please downgrade to `log_warning_errno()`, and drop the `return`

poettering commented on src/network/networkd-link.c:
```diff
@@ -1829,6 +1872,14 @@ void link_drop(Link *link) {
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
+        }
+        hashmap_remove_value(link->manager->links_by_name, link->ifname, link);
```
please move the name unregistration into a function of its own, too, also covering both the alternative names and the main name

poettering commented on src/network/networkd-link.c:
```diff
@@ -2714,8 +2765,18 @@ int link_update(Link *link, sd_netlink_message *m) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
```
if you have that unregistration function i suggested above, you could just call it here, simplifying things quite a bit

poettering commented on src/network/networkd-link.c:
```diff
@@ -2714,8 +2765,18 @@ int link_update(Link *link, sd_netlink_message *m) {
+                        return log_error_errno(r, "Failed to add alternative name '%s' on network interface '%s', ignoring: %m", *n, link->ifname);
+
                 strv_free_and_replace(link->alternative_names, s);
+        }
```
the main name of an iface can chance too, and needs to be covered. if you properly encapsulate the update/removal functions as suggested, then you will get this for free, as one will remove the whole shebang including main name, and the other will register the whole shebang

ssahani replied:
I did not got this.

poettering replied:
the main name of the interface can change anytime too, and we should cover that here too and update the by-name hash tables, too

poettering commented on src/network/networkd-manager.h:
```diff
@@ -41,6 +41,7 @@ struct Manager {
         LinkAddressState address_state;
 
         Hashmap *links;
+        Hashmap *links_by_name;
```
maybe add a comment to both these tlines:

```c
        Hashmap *links; /* keyed by ifindex */
        Hashmap *links_by_name; /* keyed by main + alternative interface names */
```

poettering reviewed (changes requested):
sorry for the late review. 

final review I figure we should leave to @yuwata

ssahani commented:
updated. Thanks for the review. I did not get the last point though

poettering commented on src/network/networkd-link.c:
```diff
@@ -587,6 +591,61 @@ int link_get(Manager *m, int ifindex, Link **ret) {
+        assert(link);
+
+        if (hashmap_size(link->manager->links_by_name) <= 0)
+                return 0;
```
this check looks unnecessar, should work the same if these two lines are dropped

poettering commented on src/network/networkd-link.c:
```diff
@@ -1829,6 +1888,9 @@ void link_drop(Link *link) {
         log_link_debug(link, "Link removed");
 
+        if (link->alternative_names)
+                (void) link_remove_by_name(link);
```
the conditionalization should go, no? this should be called in any case

poettering reviewed (changes requested):
it would be great to split this in two commits btw: one that adds the by_name hash table, and the other one that then adds the fdb stuff based on it

ssahani commented:
updated . Thanks for the review. ptal

poettering commented on src/network/networkd-link.c:
```diff
@@ -583,6 +587,65 @@ int link_get(Manager *m, int ifindex, Link **ret) {
+                hashmap_remove_value(link->manager->links_by_name, link->ifname, link);
+
+        return 0;
+}
```
maybe have two distinct functions link_remove_by_altnames() and link_remove_by_ifname() here, and then call either one or the other, or both. the boolean thing is a bit weird

poettering commented on src/network/networkd-link.c:
```diff
@@ -2664,12 +2729,25 @@ int link_update(Link *link, sd_netlink_message *m) {
                 if (r < 0)
                         return r;
+
+                (void) link_remove_by_name(link, true, false);
```
this doesn't look right, since you remove the entries for the new name here, but you should remove them for the *old* name

ssahani replied:
yeah right. I moved it up before link_drop.

poettering commented on src/network/networkd-link.c:
```diff
@@ -2664,12 +2729,25 @@ int link_update(Link *link, sd_netlink_message *m) {
+        if (r >= 0) {
                 strv_free_and_replace(link->alternative_names, s);
 
+                (void) link_remove_by_name(link, false, true);
```
same here, you remove the entries for the new nmes, not the old names

poettering commented:
the second commit looks good

ssahani commented:
Updated. Thanks for the patience.

awlx commented:
Any updates on this PR? That's also a feature we would need :).

poettering commented:
PR looks good to me, but @yuwata should have a final review I guess. Also, needs a rebase.
